app/product_notices.go

```diff
@@ -0,0 +1,260 @@
+		if err != nil {
+			return false, errors.Wrapf(err, "Cannot parse version range %s", v)
+		}
+		if !c.Check(serverVersion) {
+			return false, nil
+		}
+	}
+
+	// check if sku matches our license
+	if cnd.Sku != nil {
+		if !cnd.Sku.Matches(sku) {
+			return false, nil
+		}
+	}
+
+	// check the target audience
+	if cnd.Audience != nil {
+		if !cnd.Audience.Matches(isSystemAdmin, isTeamAdmin) {
+			return false, nil
+		}
+
+	}
+
+	// check user count condition against previously calculated total user count
+	if cnd.NumberOfUsers != nil && userCount > 0 {
+		if userCount < *cnd.NumberOfUsers {
+			return false, nil
+		}
+	}
+
+	// check post count condition against previously calculated total post count
+	if cnd.NumberOfPosts != nil && postCount > 0 {
+		if postCount < *cnd.NumberOfPosts {
+			return false, nil
+		}
+	}
+
+	// check if our server config matches the notice
+	for k, v := range cnd.ServerConfig {
+		if !validateConfigEntry(a, k, v) {
+			return false, nil
+		}
+	}
+
+	// check the type of installation
+	if cnd.InstanceType != nil {
+		if !cnd.InstanceType.Matches(isCloud) {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func validateConfigEntry(a *App, path string, expectedValue interface{}) bool {
+	value, found := config.GetConfigValueByPath(a.Config(), strings.Split(path, "."))
+	if !found {
+		return false
+	}
+	vt := reflect.ValueOf(value)
+	if vt.IsNil() {
+		return expectedValue == nil
+	}
+	if vt.Kind() == reflect.Ptr {
+		vt = vt.Elem()
+	}
+	val := vt.Interface()
+	return val == expectedValue
+}
+
+func (a *App) GetProductNotices(lastViewed int64, userId, teamId string, client model.NoticeClientType, clientVersion string, locale string) (model.NoticeMessages, *model.AppError) {
+	isSystemAdmin := a.SessionHasPermissionTo(*a.Session(), model.PERMISSION_MANAGE_SYSTEM)
+	isTeamAdmin := a.SessionHasPermissionToTeam(*a.Session(), teamId, model.PERMISSION_MANAGE_TEAM)
+
+	// check if notices for regular users are disabled
+	if !*a.Srv().Config().AnnouncementSettings.UserNoticesEnabled && !isTeamAdmin && !isSystemAdmin {
+		return []model.NoticeMessage{}, nil
+	}
+
+	// check if notices for admins are disabled
+	if !*a.Srv().Config().AnnouncementSettings.AdminNoticesEnabled && (isTeamAdmin || isSystemAdmin) {
+		return []model.NoticeMessage{}, nil
+	}
+
+	views, err := a.Srv().Store.ProductNotices().GetViews(userId)
+	if err != nil {
+		return nil, model.NewAppError("GetProductNotices", "api.system.update_viewed_notices.failed", nil, err.Error(), http.StatusBadRequest)
+	}
+
+	sku := a.Srv().ClientLicense()["SkuShortName"]
+	isCloud := a.Srv().ClientLicense()["Cloud"] != ""
+
+	var filteredNotices []model.NoticeMessage
+
+	getViewState := func(nid string) *model.ProductNoticeViewState {
+		for _, v := range views {
+			if v.NoticeId == nid {
+				return &v
+			}
+		}
+		return nil
+	}
+
+	for _, notice := range cachedNotices {
+		// check if the notice has been viewed already
+		view := getViewState(notice.ID)
```

Comment:
this inline function looks a bit more complex to understad for me than a regular inline block, somethig like:

```suggestion
		var view *model.ProductNoticeViewState
		for _, v := range views {
			if v.NoticeId == nid {
				view = &v
				break
			}
		}
```

But maybe is matter of taste.